src/solver/loss/forward_smooth_l1loss.cpp

```diff
@@ -0,0 +1,132 @@
+#include <miopen/loss/solvers.hpp>
+#include <miopen/smooth_l1loss.hpp>
+#include <miopen/target_properties.hpp>
+
+#define LOCAL_SIZE 64
+
+namespace miopen {
+
+namespace solver {
+
+namespace loss {
+
+bool IsImprovementOverROCm(const ExecutionContext&  /*context*/,
+                           const miopen::loss::ProblemDescription&  /*problem*/)
+{
+    return true;
+}
+
+bool SmoothL1LossUnreducedForward::IsApplicable(
+    const ExecutionContext& context, const miopen::loss::ProblemDescription& problem) const
+{
+    if(!problem.IsSameType())
+        return false;
+    if(!problem.IsRightLength())
+        return false;
+    if(!problem.IsCorrectReduction())
+        return false;
+    if(!IsImprovementOverROCm(context, problem))
+        return false;
+    return true;
+}
+
+ConvSolution
+SmoothL1LossUnreducedForward::GetSolution(const ExecutionContext&  /*context*/,
+                                          const miopen::loss::ProblemDescription& problem) const
+{
+    auto result = ConvSolution{miopenStatusSuccess};
+
+    auto dtype = problem.GetIDesc().GetType();
+    auto size = problem.GetIDesc().GetElementSize();
+
+    {
+        size_t xlocalsize = LOCAL_SIZE;
+        size_t xgridsize  = AlignUp(size, xlocalsize);
+        size_t ylocalsize = 1;
+        size_t ygridsize  = 1;
+        size_t zlocalsize = 1;
+        size_t zgridsize  = 1;
+
+        auto kernel = KernelInfo{};
+
+        kernel.kernel_file = "MIOpenSmoothL1Loss.cpp";
+        kernel.kernel_name = "SmoothL1LossUnreducedForwardContiguous";
+
+        const auto build_params = KernelBuildParameters{
+            {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+            {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+            {"MIOPEN_USE_FP64", static_cast<int>(dtype == miopenDouble)},
+            {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+        };
+
+        kernel.comp_options = build_params.GenerateFor(kbp::HIP{});
+
+        kernel.l_wk.push_back(xlocalsize);
+        kernel.l_wk.push_back(ylocalsize);
+        kernel.l_wk.push_back(zlocalsize);
+
+        kernel.g_wk.push_back(xgridsize);
+        kernel.g_wk.push_back(ygridsize);
+        kernel.g_wk.push_back(zgridsize);
+
+        result.construction_params.push_back(kernel);
+    }
+
+    result.invoker_factory = [](const std::vector<Kernel>& kernels) {
+            return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+                decltype(auto) kernel = handle_.Run(kernels.front());
+                decltype(auto) params = raw_params.CastTo<miopen::loss::InvokeParams>();
+
+                auto size = params.iDesc->GetElementSize();
+
+                kernel(params.i,
+                       params.t,
+                       params.o,
+                       params.beta,
+                       size);
+            };
+        };
+
+    return result;
+}
+
+std::size_t SmoothL1LossUnreducedForward::GetWorkspaceSize(
```

Comment:
This part appears to be currently unused, too.

Reply:
Well, for now, it is unused, but it is needed for implementing the reduction version of SmoothL1Loss. I will keep this function and those which are related to Smooth L1Loss memory management.